compositor_pipeline/src/pipeline/encoder/fdk_aac.rs

```diff
@@ -0,0 +1,354 @@
+            } else {
+                break;
+            }
+        }
+        // Encode should be called after `enqueue_input`, so unwrap is safe.
+        let pts = self.start_pts.unwrap()
+            + Duration::from_secs_f64(self.sent_samples as f64 / self.sample_rate as f64);
+        self.sent_samples += samples_encoded;
+
+        match output.is_empty() {
+            true => Ok(None),
+            false => Ok(Some(EncodedChunk {
+                data: output.freeze(),
+                pts,
+                dts: None,
+                kind: EncodedChunkKind::Audio(AudioCodec::Aac),
+            })),
+        }
+    }
+
+    fn enqueue_samples(&mut self, samples: OutputSamples) {
+        if self.start_pts.is_none() {
+            self.start_pts = Some(samples.start_pts);
+        };
+
+        match samples.samples {
+            AudioSamples::Mono(mono_samples) => {
+                self.input_buffer.extend(mono_samples.iter());
+            }
+            AudioSamples::Stereo(stereo_samples) => {
+                for (l, r) in stereo_samples {
+                    self.input_buffer.push(l);
+                    self.input_buffer.push(r);
+                }
+            }
+        }
+    }
+}
+
+impl Drop for AacEncoderInner {
+    fn drop(&mut self) {
+        unsafe {
+            fdk::aacEncClose(&mut self.encoder as *mut _);
+        }
+    }
+}
+
+fn run_encoder_thread(
+    init_result_sender: Sender<Result<(), fdk::AACENC_ERROR>>,
+    options: AacEncoderOptions,
+    sample_rate: u32,
+    samples_batch_receiver: Receiver<PipelineEvent<OutputSamples>>,
+    packets_sender: Sender<EncoderOutputEvent>,
+) {
+    let mut encoder = match AacEncoderInner::new(options, sample_rate) {
+        Ok(encoder) => {
+            init_result_sender.send(Ok(())).unwrap();
+            encoder
+        }
+        Err(err) => {
+            init_result_sender.send(Err(err)).unwrap();
+            return;
+        }
+    };
+
+    let send_encode_res = |res: Result<Option<EncodedChunk>, fdk::AACENC_ERROR>| match res {
+        Ok(Some(encoded_samples)) => {
+            if packets_sender
+                .send(EncoderOutputEvent::Data(encoded_samples))
+                .is_err()
+            {
+                info!("Failed to send AAC encoded samples.");
+            };
+        }
+        Ok(None) => {}
+        Err(err) => {
+            error!("Error encoding audio samples: {:?}", err);
+        }
+    };
+
+    loop {
+        match samples_batch_receiver.recv() {
+            Ok(PipelineEvent::Data(samples)) => send_encode_res(encoder.encode(samples)),
+            Err(_) | Ok(PipelineEvent::EOS) => {
+                send_encode_res(encoder.flush());
+                if packets_sender.send(EncoderOutputEvent::AudioEOS).is_err() {
+                    info!("Failed to send EOS event.");
+                };
+                break;
+            }
+        }
+    }
```

Comment:
```suggestion
     for event in samples_batch_receiver {
          let samples = match event {
             PipelineEvent::Data(samples) => samples
             PipelineEvent::EOS => break;
          };
          
          match encoder.encode(samples) {
             Ok(Some(encoded_samples)) => {
                if packets_sender
                   .send(EncoderOutputEvent::Data(encoded_samples))
                   .is_err()
               {
                    info!("Failed to send AAC encoded samples.");
                    break;
               };
             }
             Ok(None) => {}
             Err(err) => {
                 error!("Error encoding audio samples: {:?}", err);
             }
         }
     }
     
     send_encode_res(encoder.flush());
     if packets_sender.send(EncoderOutputEvent::AudioEOS).is_err() {
         info!("Failed to send EOS event.");
     };
```
Your implementation is unnecessarily hiding control flow + you are trying to send even if channel is closed